### DISCUSSION TITLE
Agents and environment implementation

### DISCUSSION THREAD

**Discussion by Rakesh-123-cryp** [mesa/mesa#2081] [Ideas]
1.  We can integrate the agents and the environment using wrapper functions and class methods like:

This below is the wrapper class

```Python
def from_externel_agent(agent_class, name):

        class AgentWrapper(mesa.Agent):
            def __init__(self, unique_id: int, model: mesa.Model, agent_class) -> None:
                super(AgentWrapper,self).__init__(unique_id, model)
                self.fgn_class = agent_class
                self.__name__ = name
                self.functions = inspect.getmembers(agent_class,predicate=inspect.ismethod)
            
          def tf_functions(self, func_name=None, atrribute=None, *args, **kwargs):
                if func_name in self.functions:
                    self.functions[func_name](*args,**kwargs)
                else:
                    raise AttributeError(f"Function {func_name} not found in {self.__name__}")
        
          def step(self):
                return self.fgn_class.step()
        
          def advance(self) -> None:
               #Add extra default setting code using the functions on tf agents
               if "advance" in self.functions:
                    return self.fgn_class.advance()
            
                return super().advance()
        
          def remove(self) -> None:
                #Add extra default setting code using the functions on tf agents
                if "remove" in self.functions:
                    return self.fgn_class.advance()
            
                return super().remove()
```

This class can be used for any tensor flow or KerasRL agents used and can be also considered for OpenAI Baselines.

2.  Where as for the Environment of the agent we can either let the user integrate the gym environment and project it on the web     based visualiser and we can also provide the user with mesa's inbuilt functionality to create environments as such which can also relatively lower the memory usage and possibly speedup the visualisation. When projecting the OpenAI environment onto the visualiser we can also provide them with hyper-parameter tuner to observe the effect in the environment.

3.  We can also create an experience replay buffer library that can hold the episodes of actions, observation, reward and the respective metrics at the end of each episode such as success rate, avg reward, cumulative reward, Short term and Long term Risks, etc.

**Comment by EwoutH** [comment]
Thanks for reaching out! I see some nice ideas, but I'm missing a bit of context here. Is this part of GSoC? What problem do you try to solve? Do you have specific questions or ideas you want to discuss?
